sys/kern/lockdep.c

```diff
@@ -0,0 +1,257 @@
+#include <sys/mimiker.h>
+#include <sys/klog.h>
+
+typedef struct lock_class {
+  SIMPLEQ_ENTRY(lock_class) hash_entry;
+  lock_class_key_t *key;
+  const char *name;
+
+  /* A list of lock classes corresponding to locks acquired while holding a lock
+   * belonging to this lock class. */
+  SIMPLEQ_HEAD(, lock_class_link) locked_after;
+
+  /* Used in the graph traversal to decide if the graph node was already
+   * visited. */
+  int bfs_gen_id;
+} lock_class_t;
+
+/*
+ * lock_class_link represents an edge in the graph of lock depedency. An edge
+ * C1-> C2 is created if the lock class C1 acquired while holding the class C2.
+ */
+typedef struct lock_class_link {
+  SIMPLEQ_ENTRY(lock_class_link) entry;
+  lock_class_t *to;
+} lock_class_link_t;
+
+static spin_t main_lock = SPIN_INITIALIZER(0);
+
+#define CLASSHASH_SIZE 64
+#define CLASSHASH(key) ((uintptr_t)(key) % CLASSHASH_SIZE)
+#define CLASS_HASH_CHAIN(key) (&lock_hashtbl[CLASSHASH(key)])
+
+static SIMPLEQ_HEAD(, lock_class) lock_hashtbl[CLASSHASH_SIZE];
+
+#define MAX_CLASSES 64
+static lock_class_t lock_classes[MAX_CLASSES];
+static int class_cnt = 0;
+
+#define MAX_LINKS 256
+static lock_class_link_t lock_links[MAX_LINKS];
+static int link_cnt = 0;
+
+#define BFS_QUEUE_SIZE 32
+
+typedef struct {
+  lock_class_link_t *items[BFS_QUEUE_SIZE];
+  size_t head, tail;
+} bfs_queue_t;
+
+static int bfs_cur_gen;
+static bfs_queue_t bfs_queue;
+
+static inline void bfs_q_init(bfs_queue_t *q) {
+  q->head = q->tail = 0;
+}
+
+static inline int bfs_q_empty(bfs_queue_t *q) {
+  return q->head == q->tail;
+}
+
+static inline int bfs_q_full(bfs_queue_t *q) {
+  return ((q->tail + 1) % BFS_QUEUE_SIZE) == q->head;
+}
+
+static inline int bfs_q_enqueue(bfs_queue_t *q, lock_class_link_t *elem) {
+  if (bfs_q_full(q))
+    return 0;
+
+  q->items[q->tail] = elem;
+  q->tail = (q->tail + 1) % BFS_QUEUE_SIZE;
+  return 1;
+}
+
+static inline lock_class_link_t *bfs_q_dequeue(bfs_queue_t *q) {
+  lock_class_link_t *elem;
+
+  if (bfs_q_empty(q))
+    return NULL;
+
+  elem = q->items[q->head];
+  q->head = (q->head + 1) % BFS_QUEUE_SIZE;
+  return elem;
+}
+
+static inline void lockdep_lock(void) {
+  spin_lock(&main_lock);
+}
+
+static inline void lockdep_unlock(void) {
+  spin_unlock(&main_lock);
+}
+
+static inline lock_class_link_t *bfs_next_link(lock_class_link_t *link) {
+  if (link == NULL)
+    return NULL;
+  return SIMPLEQ_NEXT(link, entry);
+}
+
+static inline void bfs_mark_visited(lock_class_t *class) {
+  class->bfs_gen_id = bfs_cur_gen;
+}
+
+static inline int bfs_is_visited(lock_class_t *class) {
+  return class->bfs_gen_id == bfs_cur_gen;
+}
+
+/*
+ * Check that the dependency graph starting at <src> can lead to
+ * <target> or not.
+ */
+static bool check_path(lock_class_t *src, lock_class_t *target) {
+  lock_class_link_t *child, *link = NULL;
+  bfs_queue_t *queue = &bfs_queue;
+
+  bfs_q_init(queue);
+  bfs_cur_gen++;
+
+  bfs_q_enqueue(queue, SIMPLEQ_FIRST(&src->locked_after));
```

Comment:
Is it guaranteed that the `src->locked_after` list is not empty? It's not obvious to me.

Reply:
There's no such guarantee, but it doesn't affect the algorithm. In this case we simply won't enter the loop
```
while ((link = bfs_next_link(link)) || (link = bfs_q_dequeue(queue))) {
```
because both conditions aren't fulfilled.

Reply:
I see it now. Thanks.